trieste/models/keras/sampler.py

```diff
@@ -80,29 +83,183 @@ def resample_trajectory(self, trajectory: TrajectoryFunction) -> TrajectoryFunct
+        with every new sample.
+
+        :param trajectory: The trajectory function to be resampled.
+        :return: The new resampled trajectory function.
+        """
+        tf.debugging.Assert(isinstance(trajectory, dropout_trajectory), [])
+        trajectory.resample()  # type: ignore
+        return trajectory
+
+# @tf.function
+class dropout_trajectory(TrajectoryFunctionClass):
+    """
+    Generate an approximate function draw (trajectory) by using the predicted means
+    of the stochastic forward passes as a trajectory. 
     """
 
-    def __init__(self, model: EnsembleModel):
+    def __init__(self, model: TrainableProbabilisticModel):
         """
         :param model: The model of the objective function.
         """
         self._model = model
-        self._network_index = tf.Variable(self._model.sample_index(1)[0])
-
-    @tf.function
-    def __call__(self, x: TensorType) -> TensorType:  # [N, 1, d] -> [N, 1]
-        """Call trajectory function."""
-        tf.debugging.assert_shapes(
-            [(x, [..., 1, None])],
-            message="This trajectory only supports batch sizes of one.",
+        self._initialized = tf.Variable(False, trainable=False)
+
+        self._batch_size = tf.Variable(
+            0, dtype=tf.int32, trainable=False
         )
-        x = tf.squeeze(x, -2)  # [N, D]
-        return self._model.predict_ensemble(x)[0][self._network_index]
+
+        self._seeds = tf.Variable(
+            tf.ones([0,0], dtype=tf.int32), shape=[None, None], trainable=False
+        )
+            
+    def __call__(self, x: TensorType) -> TensorType:  # [N, B, d] -> [N, B]
+        """
+        Call trajectory function. It uses `tf.random` seeds to fix the matrix of dropout
+        inputs or weights in the kernel, and performs a single forward pass to return the
+        equivalent to a posterior draw.
+        """
+        if not self._initialized:  # work out desired batch size from input
+            self._batch_size.assign(tf.shape(x)[-2])  # B
+            self.resample() # sample B seeds to fix
+            self._initialized.assign(True)
+
+        tf.debugging.assert_equal(
+            tf.shape(x)[-2],
+            self._batch_size.value(),
+            message=f"""
+            This trajectory only supports batch sizes of {self._batch_size}.
+            If you wish to change the batch size you must get a new trajectory
+            by calling the get_trajectory method of the trajectory sampler.
+            """
+        )
+
+        predictions = []
+        batch_index = tf.range(0, self._batch_size, 1)
+        _ = self._model.sample(x[:1,0,:], num_samples=1) # [DAV] Somehow first seed doesn't propagate unless I've done a dummy pred before.
+        for b, seed in zip(batch_index, tf.unstack(self._seeds)):
+            tf.random.set_seed(seed) # [DAV] A possible local seed? One can pass operational seeds to both types of dropouts, but it doesn't seem to fix the prediction
```

Comment:
@ampudia19 setting a global TF seed seems wrong way to go about this, its also incomplete, few more seeds need to be set to make sure things are really deterministic...

can we have a method in the model that sets/masks weights in the dropout layers where we generate this with a seed and a rate? then we could call this method for each trajectory and with seeds that you generated here, and this would make sure we are having the same trajectory?

Reply:
All deterministic tests I proposed, it passed. But it is clearly incompatible with tf.function, so a better solution is indeed necessary.

Clinton proposed a very similar solution to yours. I'll happily implement it, and will be nice to talk over it tomorrow.

Reply:
there is an oldish discussion here https://stackoverflow.com/questions/56707661/manually-assign-dropout-layer-in-keras that might be useful

Reply:
ah, this seems like exactly what we need: https://github.com/keras-team/keras/issues/14172

Reply:
(although set_weights might be computationally expensive...)

Reply:
I tried setting weights with a minimum working example, but I couldn't get_weights while in graph execution..

Reply:
Perhaps the behavior of set_weights is completely different from that of get_weights.

Reply:
another option could be to overwrite Dropout layer class and somehow allow for seed argument